src/main/kotlin/org/pkl/intellij/util/Util.kt

```diff
@@ -340,3 +344,17 @@ fun decodePath(path: String): String {
 
 fun <T> noCacheResult(): CachedValueProvider.Result<T> =
   CachedValueProvider.Result.create(null, ModificationTracker.EVER_CHANGED)
+
+fun <T, R> CompletableFuture<T>.handleOnEdt(handler: (T?, Throwable?) -> R): CompletableFuture<R> {
+  fun extractError(error: Throwable): Throwable {
+    return when (error) {
+      is CompletionException -> extractError(error.cause!!)
+      is ExecutionException -> extractError(error.cause!!)
+      else -> error
+    }
+  }
```

Comment:
[nit] move this to the top level (avoid allocating a new function every time `handleOnEdt` is called)